### PR TITLE
[Cash App Pay] Add support for splitting payments between a Square Gift Card and Cash App Pay.

### LINKED ISSUE #64: [Cash App Pay] Add support for partial payments, where a shopper might use a gift card in tandem with CashApp to fully pay for an order.
Add support for the partial payments in Cash App Pay, which enables customers to split a payment between a Square Gift and a Cash App Pay if the Gift card has insufficient funds.

### PR DESCRIPTION
**Pull request by iamdharmesh**
### All Submissions:

* [x] Does your code follow the [WooCommerce Sniffs](https://github.com/woocommerce/woocommerce-sniffs/) variant of WordPress coding standards?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?
* [ ] Will this change require new documentation or changes to existing documentation?

---

### Changes proposed in this Pull Request:
PR adds support for the split payment between Square Gift Card and Cash App Pay if the gift card has insufficient funds. Currently, we already have split payment support for Square Credit card, this PR extend this support for the Cash App Pay.

> [!NOTE]  
> 1.  Splitting payments is currently possible only on the traditional checkout (shortcode-based). As gift card support has not yet been added to the block checkout (see https://github.com/woocommerce/woocommerce-square/issues/32).
> 2. PR also contains changes from the branch [`enhancement/60`](https://github.com/woocommerce/woocommerce-square/tree/enhancement/60) to accommodate delayed capture. If PR https://github.com/woocommerce/woocommerce-square/pull/62 has not been merged yet, please compare the differences using this link: https://github.com/woocommerce/woocommerce-square/compare/enhancement/60...enhancement/64

> [!IMPORTANT]  
> This PR is not tested with the Square production environment, it should be tested with the Square production environment before we merge/release this.

Closes #64 

### Steps to test the changes in this Pull Request:
1. Add any product to the cart (The product price should be higher than the gift card value).
1. Go to Checkout (shortcode checkout).
1. Enter the gift card number (`7783 3200 0000 0000`) and click Apply. Make sure that you see a note which mentions tha you have to pay the remaining amount with a credit card and Cash App Pay.
1. Make sure the "Cash App Pay" payment gateway is available in the list of available payment gateways to pay the remaining amount of the order.
1. Complete the remaining payment with Cash App Pay and ensure the order is placed successfully.
1. Verify that the payment method value shows both payment methods with the correct amount paid on the order received page, order list page in the admin, and edit order page.
1. Verify that order transaction details (transaction amount and IDs) for both the gift card and Cash App Pay are in the admin order notes.

**Refund:**
1. Perform a full refund on an order placed with split payment between a gift card and Cash App Pay.
1. Perform a partial refund on an order placed with split payment between a gift card and Cash App Pay.
1. Perform a partial refund (custom amount) on an order placed with split payment between a gift card and Cash App Pay.
    - If the refunded amount is more than the amount charged on the gift card, only the charged amount should be refunded to the gift card. The remaining refund amount should be refunded to the cash app pay.
    - If the refunded amount is less than the amount charged on the gift card, the refund should be first done to the gift card.
    - In the case of multiple refunds using custom amounts, the amount should be refunded to the gift card until the entire amount charged on the gift card is refunded.
    - If there are more refunds (custom amount) but the gift card amount is fully refunded, then further refunds should be done to the cash app pay.

**Delayed Capture:**
1. Go to Cash App Pay settings (WooCommerce > Settings > Payments > Cash App Pay (Square)).
1. Update the transaction type to "Authorization".
1. Place an order with split payment between a Square gift card and Cash App Pay (Follow the steps given above).
1. Verify the order status is On-Hold and the payment isn't captured.
1. Go to edit order and click Capture Charge.
1. Verify the order charge gets captured successfully.

### CODE REVIEW

File: includes/Gateway/Cash_App_Pay_Gateway.php [hunk 1]
```diff
@@ -1067,6 +1152,9 @@ protected function do_transaction( $order ) {
 				$location_id = $this->get_plugin()->get_settings_handler()->get_location_id();
 				$response    = $this->get_api()->create_order( $location_id, $order );
 
+				// TODO: save gift card order details
```
Comment (dkotter) [inline includes/Gateway/Cash_App_Pay_Gateway.php]: Is this TODO still valid?

File: includes/Gateway/Gift_Card.php [hunk 2]
```diff
@@ -319,7 +319,7 @@ public function add_gift_card_fragments( $fragments ) {
 					<td><?php echo wc_price( $cart_total - $response['difference'], array( 'currency' => get_woocommerce_currency() ) ); ?></td>
 				</tr>
 				<tr>
-					<td><?php esc_html_e( 'Credit card', 'woocommerce-square' ); ?></td>
+					<td><?php esc_html_e( 'Credit card/Cash App Pay', 'woocommerce-square' ); ?></td>
```
Comment (dkotter) [inline includes/Gateway/Gift_Card.php]: Do we want to hardcode Cash App Pay here and in the line below? I guess wondering what happens if someone doesn't have Cash App Pay enabled? I'm assuming this message will still show which may be confusing.

Comment (iamdharmesh) [reply includes/Gateway/Gift_Card.php]: Nice catch! addressed in https://github.com/woocommerce/woocommerce-square/pull/70/commits/38adc7d744a5c3fa01ccec4bfc453440b5aa17dd

File: includes/Gateway/Cash_App_Pay_Gateway.php (hunk 1 above)
Comment (iamdharmesh) [reply includes/Gateway/Cash_App_Pay_Gateway.php]: This is no longer valid, as this is handled in #79 (The initial plan was to handle in this PR)

**Review by dkotter — APPROVED** [review]

**Comment by qasumitbagthariya** [comment]
### QA Update ✅ 

-------------

I have verified the Split payment in the `enhancement/64` branch and is functioning as intended.

I tested the following on this branch:

- Split payment between square gift card and cash app  ✅ 
- Refund  ✅ 
- Delayed Capture  ✅ 


**Testing Environment** 

<Details>

- WordPress: 6.4.3
- Theme: Storefront 4.5.4
- Theme: Twenty Twenty-Four 1.0
- WooCommerce - 8.6.1
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 

 </Details>

https://github.com/woocommerce/woocommerce-square/assets/67687255/84a5485d-3b77-47fe-9cf6-76d7378bf11c


https://github.com/woocommerce/woocommerce-square/assets/67687255/84d481a7-2046-4860-997c-372e47f6080f


https://github.com/woocommerce/woocommerce-square/assets/67687255/d78d6f44-3b17-44a5-bc15-38d5dd6383bd




Steps to Test- As mentioned in the PR description.
Test Results - It is working as expected.
Functional Demo / Screencast - 
Special Notes -  Ready for code review (Woo)
Testing Document status:
Cases related to this Issue/PR are added to the Critical Flow Wiki pages: 

- [ ] Yes
- [ ] Not Required/Applicable for this PR

**Review by qasumitbagthariya — APPROVED** [review]

**Comment by a-danae** [comment]
Hey team 👋  Thanks for your work here!

I'd need a hand with testing. I'm getting an error in the following step. Mind helping me see what I'm missing to test this flow?

> 3. Enter the gift card number and click Apply. (...)

I'm getting the following error when trying to use a gift card I purchased through the store:

> Enter a valid gift card number.

The request fails with a 400 code and the following data

```
{
    "errors": [
        {
            "category": "INVALID_REQUEST_ERROR",
            "code": "SANDBOX_NOT_SUPPORTED",
            "detail": "Invalid card number for sandbox.",
            "field": "giftcard_data.number"
        }
    ]
}
```

image

The gift card number I used is the one I got through email when I purchased it through the store

image

**Comment by vikrampm1** [comment]
@iamdharmesh @qasumitbagthariya can you able to help Danae here on the testing so she can complete the review? Thanks.

**Comment by iamdharmesh** [comment]
Hi @a-danae,

Thank you for checking on this, and apologies for not mentioning the gift card number in the test steps. Could you please use the gift card number `7783 3200 0000 0000` (Copied from doc [here](https://developer.squareup.com/docs/devtools/sandbox/payments)) for sandbox testing? We are using this gift card number for testing in the sandbox.

Please let me know if you need any further information.

Thank you.

**Comment by a-danae** [comment]
> Could you please use the gift card number 7783 3200 0000 0000 (Copied from doc [here](https://developer.squareup.com/docs/devtools/sandbox/payments)) for sandbox testing? We are using this gift card number for testing in the sandbox.

Thanks @qasumitbagthariya ! That worked. Would it be worth including this in the [testing instructions](https://github.com/woocommerce/woocommerce-square/wiki/Critical-flows-WooCommerce-Square) somewhere? Glad to learn if I missed it, though.

**Review by a-danae — COMMENTED** [review]
Thanks for providing further instructions! I was able to go through the described flows now and they work as expected:

When paying for the order with a gift card and Cash App Pay:

- The order gets placed successfully
- The value paid with each payment method is correctly displayed on:
    - The order received page
    - Edit order page
    - ❓ Order list page in the admin - Where should this information be displayed? 
- The transaction amount and ID are displayed in the order notes for both the gift card and Cash App Pay.
- Using delayed capture works as expected.

**Refunds**

Refunds are working as described in the PR description - The order status changed to “refunded”, the transaction IDs are logged in the order notes for both Cash App Pay and the gift card, the refund is reflected on the Square dashboard, the refunded amount is distributed as described between the payment methods. Tested scenarios:

- Doing a full refund
- Doing a partial refund of an amount greater than what was paid bia the gift card.
- Doing a partial refund of an amount less than what was paid via the gift card.
- Doing multiple partial refunds with amounts under what was paid via the gift card, until the gift card is fully refunded.


❓ There’s a strange behavior on the Orders page I can't replicate in trunk. Mind checking if you can replicate on your end?

In this branch, only some of the orders are displayed 
 

In trunk, all of the orders are displayed

**Comment by iamdharmesh** [comment]
@a-danae thank you for checking this in detail.

> ❓ Order list page in the admin - Where should this information be displayed?

It will be displayed under the "Billing" column.
image

> ❓ There’s a strange behavior on the Orders page I can't replicate in trunk. Mind checking if you can replicate on your end?

Thanks for catching this. I tried some scenarios and was able to replicate it. It has been fixed in https://github.com/woocommerce/woocommerce-square/pull/70/commits/67d5140f83bdaa4c10c6f1cdfe92d46426e6a9a5. Could you please help to verify if this fixes the issue for you.

Thank you.

**Review by a-danae — APPROVED** [review]
> I tried some scenarios and was able to replicate it. It has been fixed in https://github.com/woocommerce/woocommerce-square/commit/67d5140f83bdaa4c10c6f1cdfe92d46426e6a9a5. Could you please help to verify if this fixes the issue for you.

Thanks for the quick fix, @iamdharmesh ! I can confirm that this fixed the issue. It now works as expected.

> It will be displayed under the "Billing" column.

And I can see the information being displayed under the "Billing" column.
image

Things look good to me 👍 thanks for all the work behind this functionality!

**Comment by qasumitbagthariya** [comment]
Regression / Smoke Test Report  ✅ 
-------

Tested with Archive File created via "php woorelease.phar build repo_URL" (Composer version 2.5.5, npm version 8.19.4, node version 16.20.0)

Status- Working expected with Plugin Archive/Zip file same as fix specific branch.


**Testing Environment** 

<Details>

- WordPress: 6.4.3
- Theme: Storefront 4.5.4
- Theme: Twenty Twenty-Four 1.0
- WooCommerce - 8.7.0
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 
- Git Branch: smoke-testing

</Details>

Next Step- Ready to Merge 🚀

Status update:
@vikrampm1

**Pull request merged**